test-tests/shared/src/test/scala/zio/test/mock/MockSpec.scala

```diff
@@ -0,0 +1,350 @@
+          (Module.maxParams _).tupled(intTuple22).flip,
+          equalTo("foo")
+        ),
+        testSpec("returns failureF")(
+          MaxParams(equalTo(intTuple22)) returns failureF(i => s"foo $i"),
+          (Module.maxParams _).tupled(intTuple22).flip,
+          equalTo("foo (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22)")
+        ),
+        testSpec("returns failureM")(
+          MaxParams(equalTo(intTuple22)) returns failureM(i => IO.failNow(s"foo $i")),
+          (Module.maxParams _).tupled(intTuple22).flip,
+          equalTo("foo (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22)")
+        )
+      ),
+      suite("looped")(
+        testSpecTimeboxed("returns never")(500.millis)(
+          Looped(equalTo(1)) returns never,
+          Module.looped(1),
+          isNone
+        )
+      )
+    ),
+    suite("assertions composition")(
+      testSpec("&&")(
+        SingleParam(equalTo(3) && isWithin(1, 5)) returns valueF(input => s"foo $input"),
+        Module.singleParam(3),
+        equalTo("foo 3")
+      ),
+      testSpec("||")(
+        SingleParam(equalTo(10) || isWithin(1, 5)) returns valueF(input => s"foo $input"),
+        Module.singleParam(3),
+        equalTo("foo 3")
+      )
+    ),
+    suite("expectations failure")(
+      testSpecDied("invalid arguments")(
+        Command(equalTo(1)) returns unit,
+        Module.command(2),
+        equalTo(InvalidCallException(List(InvalidArguments(Command, 2, equalTo(1)))))
+      ),
+      testSpecDied("invalid method")(
+        Command(equalTo(1)) returns unit,
+        Module.singleParam(1),
+        equalTo(InvalidCallException(List(InvalidMethod(SingleParam, Command, equalTo(1)))))
+      ), {
+        type E0 = Chain[Module]
+        type E1 = Call[Module, Int, Unit, Unit]
+        type L  = List[Expectation[Module]]
+        type X  = UnsatisfiedExpectationsException[Module]
+
+        def cmd(n: Int) = Command(equalTo(n)) returns unit
+
+        def hasCall(index: Int, satisfied: Boolean, saturated: Boolean, invocations: List[Int]) =
+          hasAt(index)(
+            isSubtype[E1](
+              hasField[E1, Boolean]("satisfied", _.satisfied, equalTo(satisfied)) &&
+                hasField[E1, Boolean]("saturated", _.saturated, equalTo(saturated)) &&
+                hasField[E1, List[Int]]("invocations", _.invocations, equalTo(invocations))
+            )
+          )
+
+        testSpecDied("unsatisfied expectations")(
+          cmd(1) ++ cmd(2) ++ cmd(3),
+          Module.command(1),
+          isSubtype[X](
+            hasField(
+              "expectation",
+              _.expectation,
+              isSubtype[E0](
+                hasField[E0, L](
+                  "children",
+                  _.children,
+                  isSubtype[L](
+                    /**
+                     * No idea why this takes so long... @adamgfraser could you take a look?
```

Comment:
@adamgfraser Please have a look here.

Reply:
@ioleo Looking at this now.

Reply:
There is definitely something very wrong here. It looks like `Assertion#isSubtype` is being called 450,000 times in this test.

Reply:
Could this be related to the invocations search? I am running tests with multiple nested subtype assertions outside of the mocking framework and am not seeing anything unusual.

Reply:
Not really, this is an assertion on the result of the call. The Mock framework ends (in this case) with a `ZIO.die(UnsatisfiedExpectationsException(...))` and all I'm trying to do here is to check, that the structure of the failed expectation is as expected.

Reply:
Narrowing on the issue. Not sure on timing to address.

Reply:
Okay, I have a fix. Give me a minute to push.